Implement blocking=render attribute for <script> and <style>

mattwoodrow opened the pull request:
#### 184cdfcc57261e42ed198e5874c1514a90325cd2
<pre>
Implement blocking=render attribute for &lt;script&gt; and &lt;style&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=267232">https://bugs.webkit.org/show_bug.cgi?id=267232</a>
&lt;<a href="https://rdar.apple.com/121008856">rdar://121008856</a>&gt;

Reviewed by Tim Nguyen.

Implements the &apos;blocking&apos; attribute for both &lt;script&gt; and &lt;style&gt;.

For &lt;script&gt;, it makes the element render blocking (for async scripts with a src
attribute, or modules).

&lt;style&gt; elements are already always script blocking (even when not parser
inserted) - See bug 88869. These are already implicitly render blocking due to
this, so there&apos;s currently nothing else to do other than roundtrip the
attribute.

WebKit2 has an existing, related concept called layer tree freezing that
prevents rendering updates during page transitions (for top-level documents
only). This blocking is removed when the first contentful layout is completed,
or when render blocking ends. In some of the tests this means that rendering
updates start sooner than before, when render blocking starts and then ends
before we have sufficient page content to hit the &apos;contentful&apos; milestone. It&apos;s
possible, but unlikely for real web content to do the same.

To avoid changing behaviour for sites that haven&apos;t used blocking=&quot;render&quot;
explicitly, this passes information from the HTMLScriptElement to the Document
about whether the blocking was implicit or not. The Document then only overrides
the layer tree freezing if there was at least one explicit cause of render
blocking (or any of the other visual update prevented reasons).

This fixes most of the relevant tests, except for two that use script inserted
stylesheets. These are always render blocking (and script blocking) so they
Break the tests attempts to unblock rendering. Manually removing the stylesheet
components of the test results in the rest of the test passing correctly.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/blocking-idl-attr-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-async-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-async-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-defer-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-module-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-module-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-script-expected.txt:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::executePendingScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::potentiallyBlockRendering):
(WebCore::ScriptElement::unblockRendering):
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::removedFromAncestor):
(WebCore::HTMLScriptElement::attributeChanged):
(WebCore::HTMLScriptElement::blocking):
(WebCore::HTMLScriptElement::potentiallyBlockRendering):
(WebCore::HTMLScriptElement::unblockRendering):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::attributeChanged):
(WebCore::HTMLStyleElement::blocking):
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/html/HTMLStyleElement.idl:

Canonical link: <a href="https://commits.webkit.org/282319@main">https://commits.webkit.org/282319@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6bb68c4a7b6c1819ab7d457539573a116e3847a9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42166 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13414 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64930 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49853 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13698 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/9262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65879 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39204 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54410 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/35908 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/11742 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12290 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/57454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/12073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68525 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6756 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6788 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54465 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/13933 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5649 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9461 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/37986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/39066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/40177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/38808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8310,11 +8314,18 @@ bool Document::isRenderBlocked() const
+
+        // Only forcibly complete the page transition when blocking ends (and override any
+        // rendering freezing from WebKit) if the document has explictly opted-into render blocking.
+        CompletePageTransition completePageTransition = CompletePageTransition::Yes;
```
```suggestion
        auto completePageTransition = CompletePageTransition::Yes;
```

nt1m commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -98,6 +109,41 @@ void HTMLScriptElement::setText(String&& value)
+    return scriptType() == ScriptType::Classic && isParserInserted() == ParserInserted::Yes && !hasDeferAttribute() && !hasAsyncAttribute();
+}
+
+void HTMLScriptElement::potentiallyBlockRendering()
```
Link to: https://html.spec.whatwg.org/multipage/urls-and-fetching.html#potentially-render-blocking ?

nt1m commented on Source/WebCore/html/HTMLStyleElement.cpp:
```diff
@@ -101,12 +101,28 @@ void HTMLStyleElement::attributeChanged(const QualifiedName& name, const AtomStr
     }
 }
 
+DOMTokenList& HTMLStyleElement::blocking()
```
Link to: https://html.spec.whatwg.org/multipage/semantics.html#dom-style-blocking

nt1m replied:
Also is there some unification to be done between how style blocks rendering and everything else blocks rendering?

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1926,7 +1926,8 @@ void Page::updateRendering()
 
     runProcessingStep(RenderingUpdateStep::Reveal, [] (Document& document) {
-        document.reveal();
+        if (document.visibilityState() != VisibilityState::Hidden)
+            document.reveal();
```
Maybe have a follow up to figure out why  this check isn't part of `forRenderableDocument`

mattwoodrow committed:
Implement blocking=render attribute for <script> and <style>
https://bugs.webkit.org/show_bug.cgi?id=267232
<rdar://121008856>

Reviewed by Tim Nguyen.

Implements the 'blocking' attribute for both <script> and <style>.

For <script>, it makes the element render blocking (for async scripts with a src
attribute, or modules).

<style> elements are already always script blocking (even when not parser
inserted) - See bug 88869. These are already implicitly render blocking due to
this, so there's currently nothing else to do other than roundtrip the
attribute.

WebKit2 has an existing, related concept called layer tree freezing that
prevents rendering updates during page transitions (for top-level documents
only). This blocking is removed when the first contentful layout is completed,
or when render blocking ends. In some of the tests this means that rendering
updates start sooner than before, when render blocking starts and then ends
before we have sufficient page content to hit the 'contentful' milestone. It's
possible, but unlikely for real web content to do the same.

To avoid changing behaviour for sites that haven't used blocking="render"
explicitly, this passes information from the HTMLScriptElement to the Document
about whether the blocking was implicit or not. The Document then only overrides
the layer tree freezing if there was at least one explicit cause of render
blocking (or any of the other visual update prevented reasons).

This fixes most of the relevant tests, except for two that use script inserted
stylesheets. These are always render blocking (and script blocking) so they
Break the tests attempts to unblock rendering. Manually removing the stylesheet
components of the test results in the rest of the test passing correctly.

* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/blocking-idl-attr-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-async-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-async-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-defer-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/parser-inserted-module-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-inline-module-with-import-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-module-script-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/render-blocking/script-inserted-script-expected.txt:
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::prepareScript):
(WebCore::ScriptElement::executePendingScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::potentiallyBlockRendering):
(WebCore::ScriptElement::unblockRendering):
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::removedFromAncestor):
(WebCore::HTMLScriptElement::attributeChanged):
(WebCore::HTMLScriptElement::blocking):
(WebCore::HTMLScriptElement::potentiallyBlockRendering):
(WebCore::HTMLScriptElement::unblockRendering):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/HTMLScriptElement.idl:
* Source/WebCore/html/HTMLStyleElement.cpp:
(WebCore::HTMLStyleElement::attributeChanged):
(WebCore::HTMLStyleElement::blocking):
* Source/WebCore/html/HTMLStyleElement.h:
* Source/WebCore/html/HTMLStyleElement.idl:

Canonical link: https://commits.webkit.org/282319@main
